src/hotspot/share/gc/shared/weakProcessor.inline.hpp

```diff
@@ -38,52 +38,39 @@
 class BoolObjectClosure;
 class OopClosure;
 
-template<typename IsAlive>
-class CountingIsAliveClosure : public BoolObjectClosure {
-  IsAlive* _inner;
-
-  size_t _num_dead;
-  size_t _num_total;
-
-public:
-  CountingIsAliveClosure(IsAlive* cl) : _inner(cl), _num_dead(0), _num_total(0) { }
-
-  virtual bool do_object_b(oop obj) {
-    bool result = _inner->do_object_b(obj);
-    _num_dead += !result;
-    _num_total++;
-    return result;
-  }
-
-  size_t num_dead() const { return _num_dead; }
-  size_t num_total() const { return _num_total; }
-};
-
 template <typename IsAlive, typename KeepAlive>
-class CountingSkippedIsAliveClosure : public Closure {
-  CountingIsAliveClosure<IsAlive> _counting_is_alive;
+class WeakProcessor::CountingClosure : public Closure {
+  IsAlive* _is_alive;
   KeepAlive* _keep_alive;
-
-  size_t _num_skipped;
+  size_t _old_dead;
+  size_t _new_dead;
+  size_t _live;
 
 public:
-  CountingSkippedIsAliveClosure(IsAlive* is_alive, KeepAlive* keep_alive) :
-    _counting_is_alive(is_alive), _keep_alive(keep_alive), _num_skipped(0) { }
+  CountingClosure(IsAlive* is_alive, KeepAlive* keep_alive) :
+    _is_alive(is_alive),
+    _keep_alive(keep_alive),
+    _old_dead(0),
+    _new_dead(0),
+    _live(0)
+  {}
 
   void do_oop(oop* p) {
     oop obj = *p;
     if (obj == NULL) {
-      _num_skipped++;
-    } else if (_counting_is_alive.do_object_b(obj)) {
+      ++_old_dead;
+    } else if (_is_alive->do_object_b(obj)) {
       _keep_alive->do_oop(p);
+      ++_live;
     } else {
       *p = NULL;
+      ++_new_dead;
     }
   }
 
-  size_t num_dead() const { return _counting_is_alive.num_dead(); }
-  size_t num_skipped() const { return _num_skipped; }
-  size_t num_total() const { return _counting_is_alive.num_total() + num_skipped(); }
+  size_t dead() const { return _old_dead + _new_dead; }
+  size_t new_dead() const { return _new_dead; }
+  size_t processed() const { return _new_dead + _live; }
```

Comment:
Shouldn't `_old_dead` be included in `processed()`? Looking at the old `num_total()` it includes `_num_skipped` which matches what `_old_dead` counts now.

Reply:
You are right, I messed that up; I intended to maintain the existing
reporting. There's been discussion about what numbers would be useful
there; the current set seems not ideal. I thought there was an RFE for
this, but couldn't find one, so filed:
https://bugs.openjdk.java.net/browse/JDK-8256896
This differs from
https://bugs.openjdk.java.net/browse/JDK-8215709
which is about the serial WeakProcessor.